src/blog/jwt-identification-and-hot-streams/index.md

```diff
@@ -0,0 +1,172 @@
+
+- Authenticating non-web clients (e.g., mobile apps) is becoming cumbersome.
+- Rails API-only apps usually do not use tokens for web clients authentication as well (they use tokens).
+- WebSockets have no CORS support and, thus, **vulnerable to cross-site request forgery** (or [cross-site WebSocket hijacking][cross-site-ws-hijack]).
+
+Using tokens could help us to solve these problems. What's the cost? Well, we have to build everything ourselves, both server and client side (and good luck with finding a library or gem which makes it easier not harder).
+
+While I was thinking on how to incorporate tokens into AnyCable, I realized that we can also leverage this feature to improve the performance: tokens could also carry **identification** information, and thus, could be used instead of calling an RPC server (`Authenticate` method).
+
+Let's recall how Action Cable (and AnyCable) works in the context of authentication and identification.
+
+Whenever a new connection opens, the `ApplicationCable::Connection#connect` method is called. That's the place where you can reject the connection ([`#reject_unathorized_connection`][reject-connection]) and where you configure so-called _connection identifiers_ ([`.identifed_by`][ac-identifiers]). Identifiers represent the client state which could be used by channels (e.g., `current_user` or `current_tenant`). AnyCable obtains identifiers during the `Authenticate` call and pass along all subsequent requests (in a [serialized form][identifiers-serialization]).
+
+To sum up, all we need from the `#connect` method is to accept-or-reject connection and populate identifiers. And we can encapsulate this logic within a single [JWT][] token!
+
+This is how it could look in pure Ruby. First, let's create a helper to build a connection url with a token:
+
+```ruby
+class ApplicationHelper
+  def action_cable_with_jwt_meta_tag(**identifiers)
+    # Token TTL (5 minutes)
+    exp = Time.now.to_i + 60*5
+    # JWT payload.
+    # The #serialize_identifiers is responsible for serializing objects into strings
+    # (AnyCable uses GlobalID for that)
+    payload = {ext: serialize_identifiers(identifiers), exp: exp}
+
+    token = JWT.encode payload, JWT_ENCRYPTION_KEY, "HS256"
+
+    base_url = ActionCable.server.config.url || ActionCable.server.config.mount_path
+    tag "meta", name: "action-cable-url", content: "#{base_url}?token=#{token}"
+  end
+end
+```
+
+Now we can add it to our layout:
+
+```erb
+<%= action_cable_with_jwt_meta_tag(current_user: current_user) %> would render:
+# => <meta name="action-cable-url" content="ws://demo.anycable.io/cable?token=eyJhbGciOiJIUzI1NiJ9....EWCEzziOx3sKyMoNzBt20a3QvhEdxJXCXaZsA-f-UzU" />
+```
+
+Finally, we need to verify the token and extract the identifiers in the `#connect` method:
```

Comment:
@palkan can we say "using" the connect method?

Reply:
The `#connect` method is always present; it's what's inside that matters. So, I think, "in" (or "at", or whatever).

Reply:
OK